File: .github/skills/implement-backend/patterns.cs
/*
 * patterns.java — Patrones de referencia para el agente Backend Developer en Spring Boot 3.x / Java 17+
 * Este archivo NO se ejecuta directamente. Es una referencia para que el agente
 * genere código consistente con la arquitectura Hexagonal del proyecto.
 * 
 * Uso: Cuando generes una feature nueva, sigue estos patrones para cada capa.
 */

// ─── DOMAIN LAYER: VALUE OBJECT ──────────────────────────────────────────────
package com.example.banking.domain.model;

/**
 * Value Object inmutable — encapsula lógica de un concepto de dominio.
 */
public class Email {
    private final String value;
    
    private Email(String value) {
        this.value = value;
    }
    
    public static Email create(String value) {
        if (value == null || value.isBlank() || !value.contains("@")) {
            throw new InvalidEmailException("Email format is invalid");
        }
        return new Email(value);
    }
    
    public String getValue() {
        return value;
    }
    
    @Override
    public String toString() {
        return value;
    }
}

// ─── DOMAIN LAYER: ENTITY (JPA) ───────────────────────────────────────────
package com.example.banking.domain.model;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Entidad JPA — contiene toda la lógica de cambios y validaciones de negocio.
 * - Constructor protegido para JPA
 * - Propiedades mutables solo via métodos de dominio
 * - Version para concurrencia optimista
 */
@Entity
@Table(name = "features", schema = "banking_service")
public class Feature {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(name = "uid", nullable = false, unique = true)
    private String uid;
    
    @Column(name = "name", nullable = false, length = 100)
    private String name;
    
    @Column(name = "description", length = 500)
    private String description;
    
    @Column(name = "is_archived")
    private boolean archived;
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;
    
    @Version
    @Column(name = "version")
    private Integer version;
    
    @Transient
    private List<DomainEvent> domainEvents = new ArrayList<>();
    
    protected Feature() {
        // Para JPA
    }
    
    public Feature(String uid, String name, String description) {
        this.uid = uid;
        this.name = name;
        this.description = description;
        this.archived = false;
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        this.version = 0;
        raiseDomainEvent(new FeatureCreatedEvent(uid, name));
    }
    
    // Getters
    public Long getId() { return id; }
    public String getUid() { return uid; }
    public String getName() { return name; }
    public String getDescription() { return description; }
    public boolean isArchived() { return archived; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public Integer getVersion() { return version; }
    
    // Métodos de negocio
    public void updateDetails(String name, String description) {
        if (name == null || name.isBlank()) {
            throw new InvalidFeatureDetailsException("Name cannot be empty");
        }
        this.name = name;
        this.description = description;
        this.updatedAt = LocalDateTime.now();
        this.version++;
        raiseDomainEvent(new FeatureUpdatedEvent(uid, name));
    }
    
    public void archive() {
        if (archived) {
            throw new FeatureAlreadyArchivedException("Feature is already archived");
        }
        this.archived = true;
        this.updatedAt = LocalDateTime.now();
        this.version++;
        raiseDomainEvent(new FeatureArchivedEvent(uid));
    }
    
    // Métodos privados de dominio
    private void raiseDomainEvent(DomainEvent event) {
        domainEvents.add(event);
    }
    
    public List<DomainEvent> getDomainEvents() {
        return Collections.unmodifiableList(domainEvents);
    }
    
    public void clearDomainEvents() {
        domainEvents.clear();
    }
}

// ─── DOMAIN LAYER: EXCEPTIONS ───────────────────────────────────────────────
package com.example.banking.domain.exception;

public abstract class DomainException extends RuntimeException {
    public DomainException(String message) {
        super(message);
    }
}

public class InvalidFeatureDetailsException extends DomainException {
    public InvalidFeatureDetailsException(String message) {
        super(message);
    }
}

public class FeatureAlreadyArchivedException extends DomainException {
    public FeatureAlreadyArchivedException(String message) {
        super(message);
    }
}

public class InvalidEmailException extends DomainException {
    public InvalidEmailException(String message) {
        super(message);
    }
}

// ─── DOMAIN LAYER: DOMAIN EVENTS ────────────────────────────────────────────
package com.example.banking.domain.event;

import java.time.LocalDateTime;

public interface DomainEvent {
    LocalDateTime occurredAt();
    String aggregateId();
}

public record FeatureCreatedEvent(String featureId, String name) implements DomainEvent {
    @Override
    public LocalDateTime occurredAt() {
        return LocalDateTime.now();
    }
    
    @Override
    public String aggregateId() {
        return featureId;
    }
}

public record FeatureUpdatedEvent(String featureId, String newName) implements DomainEvent {
    @Override
    public LocalDateTime occurredAt() {
        return LocalDateTime.now();
    }
    
    @Override
    public String aggregateId() {
        return featureId;
    }
}

public record FeatureArchivedEvent(String featureId) implements DomainEvent {
    @Override
    public LocalDateTime occurredAt() {
        return LocalDateTime.now();
    }
    
    @Override
    public String aggregateId() {
        return featureId;
    }
}

// ─── DOMAIN LAYER: REPOSITORY PORT (OUTPUT) ────────────────────────────────
package com.example.banking.domain.ports.out;

import com.example.banking.domain.model.Feature;
import java.util.List;
import java.util.Optional;

/**
 * Puerto de salida — Interfaz de persistencia sin JPA.
 * Es la especificación que la capa de infraestructura debe implementar.
 */
public interface FeatureRepositoryPort {
    Optional<Feature> findByUid(String uid);
    List<Feature> findAll();
    Feature save(Feature feature);
    void delete(Feature feature);
}

// ─── DOMAIN LAYER: USE CASE PORT (INPUT) ──────────────────────────────────
package com.example.banking.domain.ports.in;

import com.example.banking.application.dto.CreateFeatureDto;
import com.example.banking.application.dto.FeatureResponseDto;
import com.example.banking.application.dto.UpdateFeatureDto;
import java.util.List;

/**
 * Puerto de entrada — Casos de uso del dominio.
 */
public interface FeatureUseCase {
    FeatureResponseDto createFeature(CreateFeatureDto dto);
    FeatureResponseDto getFeatureByUid(String uid);
    List<FeatureResponseDto> getAllFeatures();
    FeatureResponseDto updateFeature(String uid, UpdateFeatureDto dto);
    void archiveFeature(String uid);
}

// ─── APPLICATION LAYER: DTOs ────────────────────────────────────────────────
package com.example.banking.application.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;

/**
 * DTOs como records para inmutabilidad.
 */
public record CreateFeatureDto(
    @NotBlank String uid,
    @NotBlank @Size(max = 100) String name,
    @Size(max = 500) String description
) {}

public record UpdateFeatureDto(
    @NotBlank @Size(max = 100) String name,
    @Size(max = 500) String description
) {}

public record FeatureResponseDto(
    Long id,
    String uid,
    String name,
    String description,
    boolean isArchived,
    LocalDateTime createdAt,
    Integer version
) {}

// ─── APPLICATION LAYER: MAPPER (MapStruct) ─────────────────────────────────
package com.example.banking.application.mapper;

import com.example.banking.domain.model.Feature;
import com.example.banking.application.dto.CreateFeatureDto;
import com.example.banking.application.dto.FeatureResponseDto;
import com.example.banking.application.dto.UpdateFeatureDto;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public interface FeatureMapper {
    
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "version", ignore = true)
    Feature toEntity(CreateFeatureDto dto);
    
    FeatureResponseDto toResponseDto(Feature entity);
    
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "uid", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "version", ignore = true)
    void updateEntityFromDto(UpdateFeatureDto dto, Feature entity);
}

// ─── DOMAIN LAYER: SERVICE (USE CASE IMPLEMENTATION) ───────────────────────
package com.example.banking.domain.service;

import com.example.banking.domain.model.Feature;
import com.example.banking.domain.ports.in.FeatureUseCase;
import com.example.banking.domain.ports.out.FeatureRepositoryPort;
import com.example.banking.application.mapper.FeatureMapper;
import com.example.banking.application.dto.CreateFeatureDto;
import com.example.banking.application.dto.FeatureResponseDto;
import com.example.banking.application.dto.UpdateFeatureDto;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Servicio de dominio — lógica de orquestación para casos de uso.
 * - Instancia el agregado
 * - Persiste cambios
 * - Publica eventos
 * - Convierte a DTO
 */
@Service
public class FeatureService implements FeatureUseCase {
    
    private final FeatureRepositoryPort repository;
    private final FeatureMapper mapper;
    
    public FeatureService(FeatureRepositoryPort repository, FeatureMapper mapper) {
        this.repository = repository;
        this.mapper = mapper;
    }
    
    @Override
    public FeatureResponseDto createFeature(CreateFeatureDto dto) {
        Feature feature = mapper.toEntity(dto);
        Feature saved = repository.save(feature);
        publishDomainEvents(saved);
        return mapper.toResponseDto(saved);
    }
    
    @Override
    public FeatureResponseDto getFeatureByUid(String uid) {
        Feature feature = repository.findByUid(uid)
            .orElseThrow(() -> new IllegalArgumentException("Feature not found: " + uid));
        return mapper.toResponseDto(feature);
    }
    
    @Override
    public List<FeatureResponseDto> getAllFeatures() {
        return repository.findAll().stream()
            .map(mapper::toResponseDto)
            .collect(Collectors.toList());
    }
    
    @Override
    public FeatureResponseDto updateFeature(String uid, UpdateFeatureDto dto) {
        Feature feature = repository.findByUid(uid)
            .orElseThrow(() -> new IllegalArgumentException("Feature not found: " + uid));
        
        mapper.updateEntityFromDto(dto, feature);
        feature.updateDetails(dto.name(), dto.description());
        
        Feature saved = repository.save(feature);
        publishDomainEvents(saved);
        return mapper.toResponseDto(saved);
    }
    
    @Override
    public void archiveFeature(String uid) {
        Feature feature = repository.findByUid(uid)
            .orElseThrow(() -> new IllegalArgumentException("Feature not found: " + uid));
        
        feature.archive();
        repository.save(feature);
        publishDomainEvents(feature);
    }
    
    private void publishDomainEvents(Feature feature) {
        // Aquí se publicarían los eventos via ApplicationEventPublisher o mensajería
        feature.clearDomainEvents();
    }
}

// ─── INFRASTRUCTURE LAYER: JPA REPOSITORY ──────────────────────────────────
package com.example.banking.infrastructure.output;

import com.example.banking.domain.model.Feature;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import java.util.Optional;

@Repository
public interface FeatureJpaRepository extends JpaRepository<Feature, Long> {
    Optional<Feature> findByUid(String uid);
}

// ─── INFRASTRUCTURE LAYER: REPOSITORY ADAPTER ──────────────────────────────
package com.example.banking.infrastructure.output;

import com.example.banking.domain.model.Feature;
import com.example.banking.domain.ports.out.FeatureRepositoryPort;
import org.springframework.stereotype.Repository;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Repository
public class FeatureRepositoryAdapter implements FeatureRepositoryPort {
    
    private final FeatureJpaRepository jpaRepository;
    
    public FeatureRepositoryAdapter(FeatureJpaRepository jpaRepository) {
        this.jpaRepository = jpaRepository;
    }
    
    @Override
    public Optional<Feature> findByUid(String uid) {
        return jpaRepository.findByUid(uid);
    }
    
    @Override
    public List<Feature> findAll() {
        return jpaRepository.findAll().stream()
            .collect(Collectors.toList());
    }
    
    @Override
    public Feature save(Feature feature) {
        return jpaRepository.save(feature);
    }
    
    @Override
    public void delete(Feature feature) {
        jpaRepository.delete(feature);
    }
}

// ─── INFRASTRUCTURE LAYER: REST CONTROLLER ────────────────────────────────
package com.example.banking.infrastructure.input;

import com.example.banking.domain.ports.in.FeatureUseCase;
import com.example.banking.application.dto.CreateFeatureDto;
import com.example.banking.application.dto.FeatureResponseDto;
import com.example.banking.application.dto.UpdateFeatureDto;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping("/api/v1/features")
public class FeatureController {
    
    private final FeatureUseCase featureUseCase;
    
    public FeatureController(FeatureUseCase featureUseCase) {
        this.featureUseCase = featureUseCase;
    }
    
    @PostMapping
    public ResponseEntity<FeatureResponseDto> createFeature(@Valid @RequestBody CreateFeatureDto dto) {
        FeatureResponseDto response = featureUseCase.createFeature(dto);
        return ResponseEntity.status(201).body(response);
    }
    
    @GetMapping("/{uid}")
    public ResponseEntity<FeatureResponseDto> getFeatureByUid(@PathVariable String uid) {
        return ResponseEntity.ok(featureUseCase.getFeatureByUid(uid));
    }
    
    @GetMapping
    public ResponseEntity<List<FeatureResponseDto>> getAllFeatures() {
        return ResponseEntity.ok(featureUseCase.getAllFeatures());
    }
    
    @PutMapping("/{uid}")
    public ResponseEntity<FeatureResponseDto> updateFeature(
        @PathVariable String uid,
        @Valid @RequestBody UpdateFeatureDto dto
    ) {
        return ResponseEntity.ok(featureUseCase.updateFeature(uid, dto));
    }
    
    @DeleteMapping("/{uid}")
    public ResponseEntity<Void> archiveFeature(@PathVariable String uid) {
        featureUseCase.archiveFeature(uid);
        return ResponseEntity.noContent().build();
    }
}

// ─── INFRASTRUCTURE LAYER: CONFIGURATION ─────────────────────────────────
package com.example.banking.infrastructure.config;

import com.example.banking.domain.ports.out.FeatureRepositoryPort;
import com.example.banking.infrastructure.output.FeatureJpaRepository;
import com.example.banking.infrastructure.output.FeatureRepositoryAdapter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class AppConfig {
    
    @Bean
    public FeatureRepositoryPort featureRepositoryPort(FeatureJpaRepository jpaRepository) {
        return new FeatureRepositoryAdapter(jpaRepository);
    }
}

File: .github/skills/implement-backend/patterns.java
/*
 * patterns.java — Patrones de referencia AGNÓSTICOS para Backend Developer
 * Spring Boot 4 / Java 21 con Arquitectura Hexagonal PURA
 * 
 * 📌 PRINCIPIO: Domain = Java PURO. Sin ningún import de framework.
 *     Las anotaciones van exclusivamente en Infrastructure y Application (para wiring).
 * 
 * Estructura Hexagonal:
 *   domain/       → Java PURO (POJO) — SIN Spring, SIN JPA, SIN nada de framework
 *   application/  → DTOs + UseCases — AQUÍ SÍ puede haber @Service (wiring)
 *   infrastructure/ → JPA Entities, Controllers, Adapters — Anotaciones de framework
 */

// ╔══════════════════════════════════════════════════════════════════════════════════════════╗
// ║                    ARQUITECTURA HEXAGONAL PURA                               ║
// ╠══════════════════════════════════════════════════════════════════════════════════════════╣
// ║                                                                          ║
// ║  DOMAIN/         → Java PURO (POJO) — 0% framework                             ║
// ║                  Solo imports: java.*, excepciones del dominio                 ║
// ║                                                                          ║
// ║  APPLICATION/   → DTOs (Records) + UseCases —aquí SÍ puede usar @Service   ║
// ║                  Solo contratos y orquestación                               ║
// ║                                                                          ║
// ║  INFRASTRUCTURE/ → JPA Entities, Controllers, Repos — 100% framework      ║
// ║                                                                          ║
// ╚══════════════════════════════════════════════════════════════════════════════════╝

package com.example.domain.model;


// ══════════════════════════════════════════════════════════════════════════════════════════════
// DOMAIN LAYER — 100% JAVA PURO (SIN NINGUNA ANOTACIÓN)
// ══════════════════════════════════════════════════════════════════════════════════════════════

// ─── VALUE OBJECT: Money (Inmutable) ──────────────────────────────────────────────────────────────
/**
 * Value Object para dinero — inmutable.
 * 
 * ⚠️ SIN anotaciones. Solo Java puro.
 */
public final class Money {
    public static final Money ZERO = new Money(java.math.BigDecimal.ZERO);
    
    private final java.math.BigDecimal amount;
    private final String currency;
    
    private Money(java.math.BigDecimal amount, String currency) {
        this.amount = amount;
        this.currency = currency;
    }
    
    public static Money of(java.math.BigDecimal amount) {
        return new Money(amount, "USD");
    }
    
    public static Money fromString(String amount) {
        return new Money(new java.math.BigDecimal(amount), "USD");
    }
    
    public Money add(Money other) {
        return new Money(this.amount.add(other.amount), this.currency);
    }
    
    public Money subtract(Money other) {
        return new Money(this.amount.subtract(other.amount), this.currency);
    }
    
    public boolean isZeroOrNegative() {
        return amount.compareTo(java.math.BigDecimal.ZERO) <= 0;
    }
    
    public boolean lessThan(Money other) {
        return this.amount.compareTo(other.amount) < 0;
    }
    
    public boolean greaterThanOrEqual(Money other) {
        return this.amount.compareTo(other.amount) >= 0;
    }
    
    public java.math.BigDecimal amount() { return amount; }
    public String currency() { return currency; }
    
    @Override public String toString() { return currency + " " + amount; }
}

// ─── VALUE OBJECT: Email ───────────────────────────────────────────────────────────────────────
/**
 * Value Object para email — inmutable.
 */
public final class Email {
    private final String value;
    
    private Email(String value) { this.value = value; }
    
    public static Email of(String value) {
        if (value == null || value.isBlank() || !value.contains("@")) {
            throw new IllegalArgumentException("Invalid email format");
        }
        return new Email(value);
    }
    
    public String value() { return value; }
    @Override public String toString() { return value; }
    @Override public boolean equals(Object o) { return o instanceof Email e && value.equals(e.value); }
    @Override public int hashCode() { return value.hashCode(); }
}

// ─── ENTITY: Account (Dominio Puro) ────────────────────────────────────────────────────────────
/**
 * Entity del dominio — lógica de negocio pura.
 * 
 * ⚠️ SIN @Entity, @Table, @Column.
 * ⚠️ Solo imports: java.*, excepciones del dominio
 */
public class Account {
    
    private Long id;
    private String accountNumber;
    private Long clientId;
    private Money balance;
    private AccountType type;
    private java.time.Instant createdAt;
    private java.time.Instant updatedAt;
    private int version;
    
    // Constructor de fábrica (única forma de crear instancias)
    private Account() {}
    
    public static Account create(String accountNumber, Long clientId, AccountType type) {
        if (accountNumber == null || accountNumber.isBlank()) {
            throw new IllegalArgumentException("accountNumber is required");
        }
        Account account = new Account();
        account.accountNumber = accountNumber;
        account.clientId = clientId;
        account.balance = Money.ZERO;
        account.type = type;
        account.createdAt = java.time.Instant.now();
        account.updatedAt = java.time.Instant.now();
        account.version = 0;
        return account;
    }
    
    // ─── Lógica de dominio (pertenece aquí, es negocio puro) ───────────────────────────
    public void deposit(Money amount) {
        if (amount == null || amount.isZeroOrNegative()) {
            throw new IllegalArgumentException("Amount must be positive");
        }
        this.balance = this.balance.add(amount);
        this.updatedAt = java.time.Instant.now();
    }
    
    public void withdraw(Money amount) {
        if (amount == null || amount.isZeroOrNegative()) {
            throw new IllegalArgumentException("Amount must be positive");
        }
        if (this.balance.lessThan(amount)) {
            throw new DomainException("Insufficient funds. Balance: " + this.balance);
        }
        this.balance = this.balance.subtract(amount);
        this.updatedAt = java.time.Instant.now();
    }
    
    public boolean hasSufficientFunds(Money amount) {
        return this.balance.greaterThanOrEqual(amount);
    }
    
    // Getters
    public Long id() { return id; }
    public String accountNumber() { return accountNumber; }
    public Long clientId() { return clientId; }
    public Money balance() { return balance; }
    public AccountType type() { return type; }
    public java.time.Instant createdAt() { return createdAt; }
    public java.time.Instant updatedAt() { return updatedAt; }
    public int version() { return version; }
    
    // Setters (para infraestructura)
    public void setId(Long id) { this.id = id; }
    public void setBalance(Money balance) { this.balance = balance; }
    public void setUpdatedAt(java.time.Instant updatedAt) { this.updatedAt = updatedAt; }
    public void setVersion(int version) { this.version = version; }
}

// ─── ENUM: AccountType ──────────────────────────────────────────────────────────────────
public enum AccountType {
    SAVINGS,
    CHECKING,
    CREDIT
}

// ─── EXCEPTIONS (Dominio Puro) ──────────────────────────────────────────────────────
public class DomainException extends RuntimeException {
    public DomainException(String message) { super(message); }
}

public class InsufficientFundsException extends DomainException {
    public InsufficientFundsException(String message) { super(message); }
}

public class AccountNotFoundException extends DomainException {
    public AccountNotFoundException(String message) { super(message); }
}


// ══════════════════════════════════════════════════════════════════════════════════════════════
// DOMAIN PORTS — INTERFACES (Java Puro, sin anotaciones)
// ══════════════════════════════════════════════════════════════════════════════════════════════

package com.example.domain.ports.in;

import com.example.domain.model.Account;
import com.example.application.dto.CreateAccountRequest;
import com.example.application.dto.AccountResponse;

/**
 * Input Port — Caso de uso del dominio.
 * 
 * ⚠️ Interfaz Java pura. Sin anotaciones de Spring.
 */
public interface AccountUseCase {
    AccountResponse createAccount(CreateAccountRequest request);
    AccountResponse getAccountById(Long id);
    AccountResponse deposit(Long id, java.math.BigDecimal amount);
    AccountResponse withdraw(Long id, java.math.BigDecimal amount);
}

package com.example.domain.ports.out;

import com.example.domain.model.Account;
import java.util.Optional;
import java.util.List;

/**
 * Output Port — Persistencia.
 * 
 * ⚠️ Interfaz Java pura. Sin JPA.
 */
public interface AccountRepository {
    Account save(Account account);
    Optional<Account> findById(Long id);
    List<Account> findByClientId(Long clientId);
    void delete(Account account);
}


// ══════════════════════════════════════════════════════════════════════════════════════════════
// APPLICATION LAYER — DTOs + USECASES (aquí SÍ puede usar @Service)
// ══════════════════════════════════════════════════════════════════════════════════════════════════════

package com.example.application.dto;

/**
 * DTOs como Records — inmutables.
 * 
 * ⚠️ SIN lógica de negocio. Solo validación de formato.
 */
public record CreateAccountRequest(
    String accountNumber,
    Long clientId,
    String accountType  // "SAVINGS", "CHECKING", "CREDIT"
) {}

public record AccountResponse(
    Long id,
    String accountNumber,
    Long clientId,
    String balance,
    String accountType,
    String createdAt,
    String updatedAt
) {}


package com.example.application.usecase;

import com.example.domain.model.Account;
import com.example.domain.model.AccountType;
import com.example.domain.model.Money;
import com.example.domain.ports.in.AccountUseCase;
import com.example.domain.ports.out.AccountRepository;
import com.example.application.dto.CreateAccountRequest;
import com.example.application.dto.AccountResponse;
import org.springframework.stereotype.Service;
import java.math.BigDecimal;
import java.util.List;

/**
 * Caso de uso — Orquesta el dominio.
 * 
 * ⚠️ AQUÍ SÍ usa @Service (es Application layer).
 *    Usa los ports para dialogar con infraestructura.
 */
@Service
public class AccountService implements AccountUseCase {
    
    private final AccountRepository repository;
    
    // Constructor injection
    public AccountService(AccountRepository repository) {
        this.repository = repository;
    }
    
    @Override
    public AccountResponse createAccount(CreateAccountRequest request) {
        AccountType type = AccountType.valueOf(request.accountType());
        Account account = Account.create(request.accountNumber(), request.clientId(), type);
        Account saved = repository.save(account);
        return toResponse(saved);
    }
    
    @Override
    public AccountResponse getAccountById(Long id) {
        Account account = repository.findById(id)
            .orElseThrow(() -> new AccountNotFoundException("Account not found: " + id));
        return toResponse(account);
    }
    
    @Override
    public AccountResponse deposit(Long id, BigDecimal amount) {
        Account account = repository.findById(id)
            .orElseThrow(() -> new AccountNotFoundException("Account not found: " + id));
        account.deposit(Money.of(amount));
        Account saved = repository.save(account);
        return toResponse(saved);
    }
    
    @Override
    public AccountResponse withdraw(Long id, BigDecimal amount) {
        Account account = repository.findById(id)
            .orElseThrow(() -> new AccountNotFoundException("Account not found: " + id));
        account.withdraw(Money.of(amount));
        Account saved = repository.save(account);
        return toResponse(saved);
    }
    
    private AccountResponse toResponse(Account account) {
        return new AccountResponse(
            account.id(),
            account.accountNumber(),
            account.clientId(),
            account.balance().toString(),
            account.type().name(),
            account.createdAt().toString(),
            account.updatedAt().toString()
        );
    }
}


// ══════════════════════════════════════════════════════════════════════════════════════════════
// INFRASTRUCTURE LAYER — JPA + SPRING (100% framework)
// ══════════════════════════════════════════════════════════════════════════════════════

package com.example.infrastructure.output;

import jakarta.persistence.*;
import java.time.Instant;

/**
 * JPA Entity — Mapea la entidad de dominio a la tabla de BD.
 * 
 * ⚠️ LAS ANOTACIONES JPA VAN AQUÍ, NO EN DOMAIN/.
 */
@Entity
@Table(name = "accounts", schema = "banking_service")
public class AccountEntity {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(name = "account_number", nullable = false, unique = true)
    private String accountNumber;
    
    @Column(name = "client_id", nullable = false)
    private Long clientId;
    
    @Column(name = "balance", nullable = false)
    private BigDecimal balance;
    
    @Column(name = "account_type", nullable = false)
    @Enumerated(EnumType.STRING)
    private String accountType;
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;
    
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;
    
    @Version
    @Column(name = "version")
    private Integer version;
    
    // Mapeo Domain → Entity
    public static AccountEntity fromDomain(Account account) {
        AccountEntity e = new AccountEntity();
        e.setId(account.id());
        e.setAccountNumber(account.accountNumber());
        e.setClientId(account.clientId());
        e.setBalance(new BigDecimal(account.balance().amount().toString()));
        e.setAccountType(account.type().name());
        e.setCreatedAt(account.createdAt());
        e.setUpdatedAt(account.updatedAt());
        return e;
    }
    
    // Mapeo Entity → Domain
    public Account toDomain() {
        Account account = Account.create(this.accountNumber, this.clientId, AccountType.valueOf(this.accountType));
        account.setId(this.id);
        account.setBalance(Money.of(this.balance));
        account.setUpdatedAt(this.updatedAt);
        account.setVersion(this.version);
        return account;
    }
    
    // Getters y Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getAccountNumber() { return accountNumber; }
    public void setAccountNumber(String accountNumber) { this.accountNumber = accountNumber; }
    public Long getClientId() { return clientId; }
    public void setClientId(Long clientId) { this.clientId = clientId; }
    public BigDecimal getBalance() { return balance; }
    public void setBalance(BigDecimal balance) { this.balance = balance; }
    public String getAccountType() { return accountType; }
    public void setAccountType(String accountType) { this.accountType = accountType; }
    public Instant getCreatedAt() { return createdAt; }
    public void setCreatedAt(Instant createdAt) { this.createdAt = createdAt; }
    public Instant getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(Instant updatedAt) { this.updatedAt = updatedAt; }
    public Integer getVersion() { return version; }
    public void setVersion(Integer version) { this.version = version; }
}

package com.example.infrastructure.output;

import com.example.domain.model.Account;
import com.example.domain.ports.out.AccountRepository;
import org.springframework.stereotype.Repository;
import java.util.Optional;
import java.util.List;

/**
 * Adapter de persistencia — Implementa el puerto de salida.
 */
@Repository
public class AccountRepositoryAdapter implements AccountRepository {
    
    private final AccountJpaRepository jpaRepository;
    
    public AccountRepositoryAdapter(AccountJpaRepository jpaRepository) {
        this.jpaRepository = jpaRepository;
    }
    
    @Override
    public Account save(Account account) {
        AccountEntity entity = AccountEntity.fromDomain(account);
        return jpaRepository.save(entity).toDomain();
    }
    
    @Override
    public Optional<Account> findById(Long id) {
        return jpaRepository.findById(id).map(AccountEntity::toDomain);
    }
    
    @Override
    public List<Account> findByClientId(Long clientId) {
        return jpaRepository.findByClientId(clientId).stream()
            .map(AccountEntity::toDomain)
            .toList();
    }
    
    @Override
    public void delete(Account account) {
        jpaRepository.delete(AccountEntity.fromDomain(account));
    }
}

package com.example.infrastructure.output;

import com.example.infrastructure.output.AccountEntity;
import org.springframework.data.jpa.repository.JpaRepository;
import java.util.List;

/**
 * Spring Data JPA Repository.
 */
@Repository
public interface AccountJpaRepository extends JpaRepository<AccountEntity, Long> {
    List<AccountEntity> findByClientId(Long clientId);
}


// ══════════════════════════════════════════════════════════════════════════════════════
// INFRASTRUCTURE LAYER — REST CONTROLLER
// ══════════════════════════════════════════════════════════════════════════════════════

package com.example.infrastructure.input;

import com.example.domain.ports.in.AccountUseCase;
import com.example.application.dto.CreateAccountRequest;
import com.example.application.dto.AccountResponse;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.time.Instant;
import java.util.Map;

/**
 * REST Controller — Adapter de entrada.
 * 
 * ⚠️ Solo routing y validación. SIN lógica de negocio.
 */
@RestController
@RequestMapping("/api/accounts")
public class AccountController {
    
    private final AccountUseCase accountUseCase;
    
    public AccountController(AccountUseCase accountUseCase) {
        this.accountUseCase = accountUseCase;
    }
    
    @PostMapping
    public ResponseEntity<?> create(@Valid @RequestBody CreateAccountRequest request) {
        AccountResponse response = accountUseCase.createAccount(request);
        return ResponseEntity.status(201).body(Map.of(
            "data", response,
            "timestamp", Instant.now().toString()
        ));
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<AccountResponse> getById(@PathVariable Long id) {
        return ResponseEntity.ok(accountUseCase.getAccountById(id));
    }
    
    @PostMapping("/{id}/deposit")
    public ResponseEntity<?> deposit(@PathVariable Long id, @RequestParam String amount) {
        AccountResponse response = accountUseCase.deposit(id, new java.math.BigDecimal(amount));
        return ResponseEntity.ok(Map.of(
            "data", response,
            "timestamp", Instant.now().toString()
        ));
    }
    
    @PostMapping("/{id}/withdraw")
    public ResponseEntity<?> withdraw(@PathVariable Long id, @RequestParam String amount) {
        AccountResponse response = accountUseCase.withdraw(id, new java.math.BigDecimal(amount));
        return ResponseEntity.ok(Map.of(
            "data", response,
            "timestamp", Instant.now().toString()
        ));
    }
}


// ══════════════════════════════════════════════════════════════════════════════════════════════
// CONFIG: Wiring (conecta Application con Infrastructure)
// ══════════════════════════════════════════════════════════════════════════════════════════════

package com.example.infrastructure.config;

import com.example.domain.ports.in.AccountUseCase;
import com.example.domain.ports.out.AccountRepository;
import com.example.application.usecase.AccountService;
import com.example.infrastructure.output.AccountJpaRepository;
import com.example.infrastructure.output.AccountRepositoryAdapter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * Configuración de wiring.
 * 
 * ⚠️Aquí se registran los beans y se conectan las capas.
 */
@Configuration
public class ApplicationConfig {
    
    @Bean
    public AccountRepository accountRepository(AccountJpaRepository jpaRepository) {
        return new AccountRepositoryAdapter(jpaRepository);
    }
    
    @Bean
    public AccountUseCase accountUseCase(AccountRepository repository) {
        return new AccountService(repository);
    }
}